server/src/main/java/org/elasticsearch/index/engine/NoOpEngine.java

```diff
@@ -116,4 +123,45 @@ public SegmentsStats segmentsStats(boolean includeSegmentFileSizes, boolean incl
             return super.segmentsStats(includeSegmentFileSizes, includeUnloadedSegments);
         }
     }
+
+    /**
+     * This implementation will trim existing translog files using a {@link TranslogDeletionPolicy}
+     * that retains nothing but the last translog generation from safe commit.
+     */
+    @Override
+    public void trimUnreferencedTranslogFiles() {
+        try (ReleasableLock lock = readLock.acquire()) {
+            ensureOpen();
+            final Map<String, String> commitUserData = getLastCommittedSegmentInfos().getUserData();
+            final String translogUuid = commitUserData.get(Translog.TRANSLOG_UUID_KEY);
+            if (translogUuid == null) {
+                throw new IllegalStateException("commit doesn't contain translog unique id");
+            }
+            if (commitUserData.containsKey(Translog.TRANSLOG_GENERATION_KEY) == false) {
+                throw new IllegalStateException("commit doesn't contain translog generation id");
+            }
+            final long lastCommitGeneration = Long.parseLong(commitUserData.get(Translog.TRANSLOG_GENERATION_KEY));
+            final TranslogConfig translogConfig = engineConfig.getTranslogConfig();
+            final long minTranslogGeneration = Translog.readMinTranslogGeneration(translogConfig.getTranslogPath(), translogUuid);
+
+            if (minTranslogGeneration < lastCommitGeneration) {
+                // a translog deletion policy that retains nothing but the last translog generation from safe commit
+                final TranslogDeletionPolicy translogDeletionPolicy = new TranslogDeletionPolicy(-1, -1);
+                translogDeletionPolicy.setTranslogGenerationOfLastCommit(lastCommitGeneration);
+                translogDeletionPolicy.setMinTranslogGenerationForRecovery(lastCommitGeneration);
```

Comment:
Should we trim translog only if we have a single index commit? We do not trim [unsafe commits](https://github.com/elastic/elasticsearch/pull/41041) when opening ReadOnlyEngine and we [might not](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java#L142) verify the condition max_seq_no = global_checkpoint if the index was created before 7.2.0. It's a bit unsafe to advance the min_translog_generation_for_recovery if we have more than one commit. WDYT?

Reply:
Thanks for spotting this. I pushed 6f08262b318bf368f6c66684ca8921aaa1eafd75, let me know what you think.